#**TTI103 - Lógica de Programação**

#**Aula L3 - Conhecendo o Python**

## **Exercício 01**
A partir da segunda lei do movimento de Newton, pode-se estabelecer uma equação matemática
que modela o movimento de uma bola lançada verticalmente para cima.

A posição vertical da bola, chamada $y$, varia com o tempo $t$ de acordo com:

$$y(t)=v_0t-\frac{1}{2}gt^2\text{   [Eq. 1]}$$

Na expressão acima, $v_0$ é a velocidade inicial da bola (em $\text{m/s}$), $g$ é a aceleração da gravidade (em $\text{m/s}^2$) e $t$ é
o tempo (em $\text{s}$).

Observe que o eixo $Oy$ é escolhido de forma que a bola esteja em $y=0$ quando $t=0$. Além disso, essa expressão despreza a resistência do ar, que geralmente é pequena, a menos que $v_0$ seja muito alta.

Para avaliar o tempo $t_r$ necessário para que a bola suba e retorne para $y=0$, podemos procurar soluções para a equação $y(t_r)=0$:

$$v_0t_r-\frac{1}{2}gt_r^2=t_r(v_0-\frac{1}{2}gt_r)=0\Rightarrow t_r=0\text{ ou }t_r=\frac{2v_0}{g}\text{   [Eq. 2]}$$

Ou seja, a bola retorna à posição inicial após $2v_0/g$ segundos. Logo, é razoável restringir nossa análise ao intervalo $t\in[0, 2v_0/g]$.

Com base nessas informações, elabore:

* Um programa Python (acompanhado de seu respectivo fluxograma) que calcule o tempo necessário para que a bola retorne à sua posição inicial, usando $\text{[Eq. 2]}$, a partir de escolhas arbitrárias de $v_0$ e $g$.

* Um programa Python (acompanhado de seu respectivo fluxograma) que calcule a posição da bola, usando $\text{[Eq. 1]}$, a partir de uma escolha de $t$ que respeite o resultado obtido no programa anterior.

In [1]:
v0 = 5
g = 9.81
tr = 2*v0/g
print(tr)

1.019367991845056


In [4]:
v0 = 5
g = 9.81
t = 0.501
y = v0*t - 0.5*g*t**2
print(y)

1.273840095


# Como escolher nomes de variáveis

* Aproveite a descrição matemática
do problema que você deseja resolver e empregue os mesmos nomes para as variáveis do programa (isso requer uma boa documentação acerca do problema!);

* Para todas as variáveis que não possuam um símbolo a elas reservado, use um nome descritivo cuidadosamente escolhido.

O programa a seguir mostra uma solução que emprega variáveis (excessivamente) descritas. O uso do bom senso é altamente recomendável!

In [5]:
velocidade_inicial = 5
aceleracao_gravidade = 9.81
tempo = 0.6
pos_vertical_bola = velocidade_inicial*tempo - 0.5*aceleracao_gravidade*tempo**2
print(pos_vertical_bola)

1.2342


# Palavras reservadas

Certas palavras são reservadas em Python porque são usadas para construir a linguagem. Estas palavras reservadas não podem ser usadas como nomes de variáveis:

`and`, `as`, `assert`, `async`, `await`, `break`, `class`, `continue`, `def`, `del`, `elif`, `else`, `except`, `False`,
`finally`, `for`, `from`, `global`, `if`, `import`, `in`, `is`, `lambda`, `None`, `nonlocal`, `not`, `or`, `pass`, `raise`, `return`, `True`, `try`, `while`, `with`, `yield`.

# Comentários

Juntamente com as declarações do programa, é aconselhável fornecer alguns comentários
em uma linguagem natural para explicar a ideia por trás das declarações.

Comentários em Python começam com o caractere `#`. Após esse caractere, todas as declarações da respectiva linha de código são ignoradas quando o programa é executado.

## **Exercício 02**

Comente os programas que você construiu no Exercício 01.

In [ ]:
''' Programa para o cálculo do tempo de retorno de uma bola lançada verticalmente para cima'''
v0 = 5    # Velocidade inicial (m/s)
g = 9.81    # Aceleração da gravidade (m/s^2)
tr = 2*v0/g    # Tempo de retorno à posição inicial (s)
print(tr)

In [ ]:
"""Programa para o cálculo do posição de uma bola lançada verticalmente para cima"""
v0 = 5 # Velocidade inicial (m/s)
g = 9.81 # Aceleração da gravidade (m/s^2)
t = 0.501 # Tempo de voo (s)
y = v0*t - 0.5*g*t**2 # Posição vertical (m)
print(y)

# Formatação de textos de saída

Existem diversos modos disponíveis no Python para formatação de textos (`strings`) de saída. Aqui, vamos explorar as `f-strings`.

Também chamadas de “*strings* literais formatadas” (formatted string literals), `f-strings` são strings que iniciam com a letra `f` e possuem chaves `{}` para realizar a interpolação das expressões que serão exibidas.

As expressões são processadas em tempo de execução e formatadas utilizadas o protocolo `__format__`.

A célula de código a seguir mostra um exemplo de aplicação. Execute-a!

In [7]:
universidade = 'Mauá'
print(f'Onde estão os melhores cursos de TI do Brasil? Na {universidade} é claro!')

Onde estão os melhores cursos de TI do Brasil? Na Mauá é claro!


Para controlar o número de casas decimais a ser exibido, basta formatar o valor da variável usando a estrutura `:.Nf`, em que:

* os dois pontos indicam que a variável à sua esquerda será formatada;
* o ponto indica o separador decimal;
* `N` é o número de casas decimais que desejamos exibir;
* `f` indicar o formato `float` para números reais (*floating-point numbers*).

*CUIDADO!!!* A letra `f` do formato `:.Nf` não tem nenhuma relação com a letra `f` das `f-strings`! No entanto, elas comumente aparecem juntas na sintaxe de exibição de informações de saída. Veja o exemplo a seguir:

In [8]:
valor_exato = 45.6732910
print(f'O resultado é {valor_exato:.2f}, com exatidão de 2 casas decimais.')

O resultado é 45.67, com exatidão de 2 casas decimais.


## **Exercício 03**

Formate as saídas dos programas que você construiu no Exercício 01. Use três casas decimais.

In [10]:
v0 = 5
g = 9.81
tr = 2*v0/g
print(f'O resultado é {tr:.2f}, com exatidão de 2 casas decimais.')

O resultado é 1.02, com exatidão de 2 casas decimais.


In [11]:
v0 = 5
g = 9.81
t = 0.501
y = v0*t - 0.5*g*t**2
print(f'O resultado é {y:.2f}, com exatidão de 2 casas decimais.')

O resultado é 1.27, com exatidão de 2 casas decimais.


# Entrada manual de dados

Para a entrada manual de dados, usaremos o comando `input`. É importante lembrar que esse comando retorna, por padrão, a informação fornecida pelo usuário no formato `string`, ou seja, um texto.

Se a entrada devesse ser entendida como um número inteiro, por exemplo, devemos converter o retorno do comando `input` para o formato `int`. Se desejamos interpretar a informação como um número real, devemos converter para `float`. No exemplo a seguir, digite $45$ e observe o resultado.

In [12]:
x = input('Digite um número inteiro: ')
print(f'Resultado = {2*x}')

Resultado = 4545


Como assim? Até ontem, $2\times 45=90$. Por que o resultado do programa foi $4545$?

Bem, o Python interpretou o valor $45$ como um texto (classe `str`).

In [15]:
print(type(x))  # type retorna o tipo da variável presente em seu argumento

<class 'str'>


Logo, quando pedimos para imprimir o dobro desse valor, o Python faz exatamente o que pedimos, repetindo o texto duas vezes!

Claro, o resultado esperado era $90$. O código a seguir mostra como devemos proceder para que o programa funcione como esperávamos. Novamente, digite o número $45$.

In [16]:
x = int(input('Digite um número inteiro: '))
print(f'Resultado = {2*x}')
print(type(x))

Resultado = 180
<class 'int'>


Ótimo! Mas e se digitarmos $3.5$? Faça isso na célula de código a seguir:

In [17]:
x = int(input('Digite um número inteiro: '))
print(f'Resultado = {2*x}')

ValueError: invalid literal for int() with base 10: '3.5'

Não deu certo... O Python nos diz:

`ValueError: invalid literal for int() with base 10: '3.5'`

Esse é um erro de valor. Pedimos que um número real fosse convertido para número inteiro, o que evidentemente não faz sentido. Assim, para que o programa funcione corretamente, devemos converter o valor digitado para `float`. Execute a célula de código abaixo, digitando $3.5$.

In [ ]:
x = float(input('Digite um número inteiro: '))
print(f'Resultado = {2*x}')
print(type(x))

## **Exercício 04**

Elabore um fluxograma e um programa em Python que calculem e exibam a média de dois números digitados pelos usuários.

* Use nomes adequados para suas variáveis;
* Comente o seu programa;
* Não se esqueça de formatar a saída! Adote duas casas decimais para a precisão da média.

In [26]:
alunoA = float(input("Digite a média do alunoA: "))
alunoB = float(input("Digite a média do alunoB: "))
média = (alunoA + alunoB)/2 
print(f'A média dos valores {alunoA} e {alunoB} é {média: .2f}')

A média dos valores 10.0 e 2.0 é  6.00
